src/ParserCombinators/Datatypes.hs

```diff
@@ -0,0 +1,101 @@
+{-# LANGUAGE GADTSyntax #-}
+{-# LANGUAGE InstanceSigs #-}
+-- {-# LANGUAGE FlexibleInstances #-}
+-- {-# LANGUAGE UndecidableInstances #-}
+
+module ParserCombinators.Datatypes where
+
+import Control.Applicative (Alternative (..))
+import Control.Monad (MonadPlus)
+
+------------------------
+---- ParseErrorType ----
+------------------------
+
+data ParseErrorType i e
+  = EndOfInput
+  | Unexpected i
+  | Expected i i
+  | ExpectedEndOfFile i
+  | CustomError e
+  | Empty
+  deriving (Eq)
+
+instance (Show i, Show e) => Show (ParseErrorType i e) where
+  show :: ParseErrorType i e -> String
+  show EndOfInput = "End of Input"
+  show (Unexpected a) = "Unexpected " ++ show a
+  show (Expected e a) = "Expected " ++ show e ++ ", got " ++ show a
+  show (ExpectedEndOfFile a) = "Expected EndOfFile, got " ++ show a
+  show (CustomError e) = "Custom Error: " ++ show e
+  show Empty = "No error"
+
+--------------------
+---- ParseError ----
+--------------------
+
+type Offset = Integer
+
+data ParseError i e where
+  ParseError
+    :: { errOffset :: Offset
+       , errError :: ParseErrorType i e
+       }
+    -> ParseError i e
+
+
+instance (Show i, Show e) => Show (ParseError i e) where
+  show :: ParseError i e -> String
+  show (ParseError o e) = "Error: " ++ show e ++ ", at offset " ++ show o
+
+----------------
+---- Parser ----
+----------------
+
+newtype Parser i m e a where
```

Comment:
Given the type of this parser, you should provide `takeWhile`, `dropWhile` (and `1` variants, returning `NonEmpty`) primitives, so that it can be a bit faster.

Also in general if you're striving for a good parser type, you should probably not be using `[]` for your token stream type, as that already rules out using `ByteString` or `Text` for your input. A better idea would be to have something like this:
```haskell
newtype Parser stream m e a where
  Parser
    :: { parse
          :: stream
          -> Offset
          -> Either
              (m (ParseError stream e))
              (Offset, a, stream)
       }
    -> Parser stream m e a
    
type Token :: Type -> Type
type family Token stream

type instance Token Text = Char

type instance Token String = Char

type instance Token ByteString = Word8
...
-- something like
token :: (Token s -> ParseErrorType stream e) -> (Token s -> Bool) -> Parser stream [] e (Token s)

```

Reply:
Thanks, this looks really promising, could you provide some learning material for type families since I'm not familiar with them

Reply:
I don't really have anything outstanding to point you to. They're almost the same thing as type level functions. I guess
https://libgen.li/edition.php?id=138121325 might be a good resource, but I don't remember anymore.

Alternatively there's https://wiki.haskell.org/GHC/Type_families#Detailed_definition_of_type_synonym_families but I never found that to be particularly enlightening.